### ISSUE TITLE
events.on_precommand broken

### ISSUE THREAD

**Issue by laerus** [xonsh/xonsh#2012]
without bashisms xontrib loaded

```
$ @events.on_precommand
. def _(*args):
.     print('pre1')
.     
$ 
pre1
pre1
$ echo okpre1
pre1
$ echo ok                                                                             
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laerus/src/xonsh/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/interface.py", line 394, in run
    self.eventloop.run(self.input, self.create_eventloop_callbacks())
  File "/usr/lib/python3/dist-packages/prompt_toolkit/eventloop/posix.py", line 164, in run
    t()
  File "/usr/lib/python3/dist-packages/prompt_toolkit/eventloop/posix.py", line 83, in read_from_stdin
    inputstream.feed(data)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/terminal/vt100_input.py", line 395, in feed
    self._input_parser.send(c)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/terminal/vt100_input.py", line 304, in _input_parser_generator
    self._call_handler(match, prefix)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/terminal/vt100_input.py", line 337, in _call_handler
    self.feed_key_callback(KeyPress(key, insert_text))
  File "/usr/lib/python3/dist-packages/prompt_toolkit/interface.py", line 1023, in feed_key
    cli.input_processor.process_keys()
  File "/usr/lib/python3/dist-packages/prompt_toolkit/key_binding/input_processor.py", line 201, in process_keys
    self._process_coroutine.send(key_press)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/key_binding/input_processor.py", line 158, in _process
    self._call_handler(matches[-1], key_sequence=buffer)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/key_binding/input_processor.py", line 228, in _call_handler
    handler.call(event)
  File "/usr/lib/python3/dist-packages/prompt_toolkit/key_binding/registry.py", line 31, in call
    return self.handler(event)
  File "/home/laerus/src/xonsh/xonsh/ptk/key_bindings.py", line 237, in multiline_carriage_return
    carriage_return(b, event.cli)
  File "/home/laerus/src/xonsh/xonsh/ptk/key_bindings.py", line 60, in carriage_return
    elif not current_line_blank and not can_compile(doc.text):
  File "/home/laerus/src/xonsh/xonsh/ptk/key_bindings.py", line 76, in can_compile
    src = src.lstrip()
AttributeError: 'NoneType' object has no attribute 'lstrip'
pre1
pre1
$ 
```

**Comment by gforsyth** [comment]
Thanks for reporting @laerus.  I can reproduce this.  I don't think that the precommand firing that's in `key_bindings.can_compile` should be there -- because the command will fire before we actually execute the command in some cases (hence the double `pre1`). 

I'm not sure we should acknowledge any code transformation that a precommand will make before deciding if we should put in a newline or execute.  That implies that we can accommodate any new syntax that anyone decides to implement via a precommand event.

**Comment by astronouth7303** [comment]
I don't think we've been planning any macros-via-events, just because of the messiness of it as soon as you start layering them.

Unfortunately, I quick perusal on GitHub only shows the even being fired in one place, so I can't be certain what's going on.

**Comment by gforsyth** [comment]
I think it's fired (for PTK) both in the `can_compile` check as part of hitting `Enter` and then again (in the normal way) in `base_shell`

**Comment by scopatz** [comment]
Hi All, I think this is because the handle that @laerus wrote does not follow [the spec](https://github.com/xonsh/xonsh/blob/master/xonsh/shell.py#L21). Namely it should return a string, not None. The following should work.

```python
@events.on_precommand
def _(cmd):
    print('pre1')
    return cmd
```

**Comment by laerus** [comment]
@scopatz allright! i should remember about the event docs.
one more strange behavior:
```
$ @events.on_precommand
. def _(cmd):
.     print("PRE1")
.     return cmd
. 
$ tt = """testPRE1     
             
. multiline"""     
. |<--- cursor
```
if i then hit enter
```
$ tt = """testPRE1
$ tt = """test
. multiline"""
. 
PRE1
$ 
```

this also happens if i'm trying to use a decorator
```
$ @events.on_chdirPRE1 

. def _(*args):
```

it seems like the handler is called twice:
```
$ from unittest.mock import Mock
$ handle = Mock(side_effect=lambda cmd: cmd)
$ events.on_precommand(handle)
<Mock id='140609579660120'>
$ handle.call_count
2
$ handle.call_count
4
$ handle.mock_calls
[call('handle.call_count\n'),
 call('handle.call_count\n'),
 call('handle.call_count\n'),
 call('handle.call_count\n'),
 call('handle.mock_calls\n'),
 call('handle.mock_calls\n')]
$ 
```

**Comment by astronouth7303** [comment]
Umm... It should mention that `on_precommand` will now fire multiple times by design. It repeatedly fires it until nobody modifies it further.

I thought that was what the macro system was for? I remember having extensive discussions and coming to the conclusion that command line modification through this becomes pretty challenging to implement in the real world.

**Comment by gforsyth** [comment]
@laerus 
Same commands on the readline shell:
```console
gil@theo ~ 🐚  xonsh --no-rc --shell-type readline
gil@theo ~ $ from unittest.mock import Mock
gil@theo ~ $ handle = Mock(side_effect=lambda cmd: cmd)
gil@theo ~ $ from xonsh.events import events
gil@theo ~ $ events.on_precommand(handle)
<Mock id='140470679610592'>
gil@theo ~ $ handle.call_count
1
gil@theo ~ $ handle.call_count
2
gil@theo ~ $ handle.mock_calls
[call('handle.call_count\n'),
 call('handle.call_count\n'),
 call('handle.mock_calls\n')]
```

This is because there's a fire precommand in `can_compile` in the ptk keybindings

**Comment by scopatz** [comment]
> This is because there's a fire precommand in can_compile in the ptk keybindings

Hi @laerus, @gforsyth is right. It is being called twice because of the `can_compile()` check in the readline shell.

> I thought that was what the macro system was for?

@astronouth7303 Yeah, but the macro system can't do arbitrary text substitutions, which is what some of the bash-like syntax does.  

>  I remember having extensive discussions and coming to the conclusion that command line modification through this becomes pretty challenging to implement in the real world.

Also agree. I don't think that this is the best implementation. IIRC, my take away was that this might should be a load event. But the way that load events are implemented now, that doesn't seem right either.  My impression is that this kind of thing is either ordered or we have to keep iterating until no changes are made. Both have advantages and disadvantages, but the main advantage right now is that ordered events don't exist :)

**Comment by scopatz** [comment]
Also, @astronouth7303, I personally think of events as your baby so I am happy to work within whatever framework you think is best. You have done way more thinking about event systems than I have!

**Comment by astronouth7303** [comment]
So, there's a huge problem with the way `on_precommand` is right now: You can't tell if it's the original or a re-run. This is essential for non-transformation use cases.

My problem with arbitrary text substitution is this: xontribs using this are expected to accurately parse and transform their syntax, even with escapes and such, without mangling or choking on other xontribs' syntaxes. Given that these syntaxes have the ability to be completely, 100% arbitrary, I consider this to be A Problem.

The current use of `on_precommand` as a transformer is [bashisms](https://github.com/xonsh/xonsh/commits/7971ee8b7ae5f2b2db69fe7ce6a16e8bc3927b35/xontrib/bashisms.py). If this use case can't be handled in a more structured way, a new event to handle the transformation should be added. (We can then also make appropriate decisions with regards to things like: Should History/loggers log the pre-transformation, post, or are offered both to log as they please?)

**Comment by scopatz** [comment]
I am all in favor of having a separate event type for a `transform_command` event!  We might be able to move other transformation events to this as well, depending on where we fire it from.  I think this will be much cleaner.

And I don't think that we can handle it in a more structured way for all things.  The bashisms in particular are weird because they don't really affect the language (in the sense that running them in a script is non-nonsensical) and they are highly coupled to the history model. However, these definitely seem to be something that some users really want. So I think that an off-by-default xontrib is a reasonable compromise.

**Comment by astronouth7303** [comment]
I'm working on this over in [event-transform_command](https://github.com/xonsh/xonsh/tree/event-transform_command), but I've realized that `on_transform_command` (as it was previously implemented) has some significant implementation flaws, namely that it only works with the ptk shell. It does not work with readline, jupyter, or scripts.

**Comment by laerus** [comment]
Hey all! Right now  returning `None` from a `on_precommand` handler makes xonsh unusable, no idea if there is way to reverse this without killing xonsh.  Maybe we can force the return value of the on_precommand to something like a 2-tuple of (pat, sub) and not actually rely on the handlers to do the parsing of the whole command. +1 @astronouth7303 for a different event for this and keep on_precommand simple.

**Comment by astronouth7303** [comment]
I'm all for discussing better ways to handle transformers. As it stands, I can see transformers being... delicate.

**Comment by scopatz** [comment]
Did you want to PR that branch @astronouth7303? I'll happily fix up the bashisms and transform handler if there is a new event to transmorgify to.

**Comment by scopatz** [comment]
Is this closable with the merge of #2063?

**Comment by laerus** [comment]
`events.on_precommand` doesn't seem to fire anymore.

**Comment by astronouth7303** [comment]
I thought I included that in #2063, but I guess it slipped through. Should be fixed in master.

**Comment by laerus** [comment]
ty @astronouth7303 !
